include/sys/fs/zfs.h

```diff
@@ -539,6 +540,17 @@ typedef enum zfs_keyformat {
 	ZFS_KEYFORMAT_FORMATS
 } zfs_keyformat_t;
 
+typedef enum zfs_passphrase_kdf {
+	ZFS_PASSPHRASE_KDF_PBKDF2 = 0,
+	ZFS_PASSPHRASE_KDF_ARGON2ID,
+	ZFS_PASSPHRASE_KDF_KDFS
+} zfs_passphrase_kdf_t;
+
+static const uint32_t zfs_passphrase_kdf_default_iterations[
+    ZFS_PASSPHRASE_KDF_KDFS] = {350000, 40000};
+static const uint32_t zfs_passphrase_kdf_min_iterations[
+    ZFS_PASSPHRASE_KDF_KDFS] = {100000, 10000};
```

Comment:
Thanks for working on this, @nabijaczleweli. I agree with @oromenahar that making the memory and parallelism parameters configurable is really important. 18 KiB seems like way too little memory. This minimum number of iterations seems way too high for Argon2; I think iteration counts as low as 1 are perfectly reasonable.

Here's are some examples from my laptop (from 2018, using <https://packages.debian.org/bookworm/argon2>).

#### t=40000 m=16KiB takes 0.3s

```console
$ echo pw | time argon2 saltsalt -id -t 40000 -m 4 -p 1 -v 13
Type:		Argon2id
Iterations:	40000
Memory:		16 KiB
Parallelism:	1
Hash:		908df7fe7b24164c854546dadce8a034fd0aa9ad04e6fed6bfb27c074757ca84
Encoded:	$argon2id$v=19$m=16,t=40000,p=1$c2FsdHNhbHQ$kI33/nskFkyFRUba3OigNP0Kqa0E5v7Wv7J8B0dXyoQ
0.349 seconds
Verification ok
argon2 saltsalt -id -t 40000 -m 4 -p 1 -v 13  0.69s user 0.00s system 99% cpu 0.691 total
```

#### t=100 m=16MiB takes 1.2s

```console
$ echo pw | time argon2 saltsalt -id -t 100 -m 14 -p 1 -v 13 
Type:		Argon2id
Iterations:	100
Memory:		16384 KiB
Parallelism:	1
Hash:		7e4127adbbe1c9ac4ec0a48f387c7c3755639d18c4f01da8b8f8c15baeee4dbc
Encoded:	$argon2id$v=19$m=16384,t=100,p=1$c2FsdHNhbHQ$fkEnrbvhyaxOwKSPOHx8N1VjnRjE8B2ouPjBW67uTbw
1.185 seconds
Verification ok
argon2 saltsalt -id -t 100 -m 14 -p 1 -v 13  2.32s user 0.02s system 99% cpu 2.349 total
```

#### t=1 m=2GiB takes 2.5s

```console
$ echo pw | time argon2 saltsalt -id -t 1 -m 21 -p 1 -v 13
Type:		Argon2id
Iterations:	1
Memory:		2097152 KiB
Parallelism:	1
Hash:		015f25635ba58191504e00830b98d57917f67e5d15a35842f108ffa1397b2b64
Encoded:	$argon2id$v=19$m=2097152,t=1,p=1$c2FsdHNhbHQ$AV8lY1ulgZFQTgCDC5jVeRf2fl0Vo1hC8Qj/oTl7K2Q
2.548 seconds
Verification ok
argon2 saltsalt -id -t 1 -m 21 -p 1 -v 13  3.76s user 1.29s system 99% cpu 5.047 total
```

This last configuration with t=1 may be best (and I should crank up parallelism if I can).

The [Argon2 paper](https://github.com/P-H-C/phc-winner-argon2/blob/master/argon2-specs.pdf) says:
> Number T of passes over the memory. The running time depends linearly on this parameter. We expect
that the user chooses this number according to the time constraints on the application. Again, there is
no ”insecure value” for T .

The recommendations here currently suggest single-digit numbers of iterations: https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#argon2id
> m=47104 (46 MiB), t=1, p=1
> ...
> m=7168 (7 MiB), t=5, p=1

Cryptsetup appears to use a minimum iteration count of 4: https://gitlab.com/cryptsetup/cryptsetup/-/blob/main/lib/crypto_backend/pbkdf_check.c#L56

It's probably also a good idea to figure out the maximum usable values and make sure that parameters coming into ZFS aren't silently truncated/wrapped.

Reply:
For me these take 0.894/2.499/4.630 (bookworm on E5645). Updated to default to 0x000100110001 (t=1, m=17, p=1) with minimum 0x0001000C0001 (t=1, m=12, p=1).